File: web/src/api/compliance.ts
import client from './client';
import type {
  AggregateComplianceMetrics,
  LineComplianceMetrics,
  DailyComplianceKPI,
} from './types';

// Compliance API functions

export const complianceApi = {
  // Get aggregate compliance metrics for all lines
  getAggregateCompliance: async (
    startDate: string,
    endDate: string,
    lineIds?: string[],
    labelIds?: string[]
  ): Promise<AggregateComplianceMetrics> => {
    const params = new URLSearchParams({
      start_date: startDate,
      end_date: endDate,
    });

    if (lineIds && lineIds.length > 0) {
      params.append('line_ids', lineIds.join(','));
    }

    if (labelIds && labelIds.length > 0) {
      params.append('label_ids', labelIds.join(','));
    }

    const response = await client.get<AggregateComplianceMetrics>(`/analytics/compliance?${params.toString()}`);
    return response.data;
  },

  // Get compliance metrics per line
  getLineCompliance: async (
    startDate: string,
    endDate: string,
    lineIds?: string[],
    labelIds?: string[]
  ): Promise<LineComplianceMetrics[]> => {
    const params = new URLSearchParams({
      start_date: startDate,
      end_date: endDate,
    });

    if (lineIds && lineIds.length > 0) {
      params.append('line_ids', lineIds.join(','));
    }

    if (labelIds && labelIds.length > 0) {
      params.append('label_ids', labelIds.join(','));
    }

    const response = await client.get<LineComplianceMetrics[]>(`/analytics/compliance/lines?${params.toString()}`);
    return response.data;
  },

  // Get daily compliance KPIs for a specific line
  getDailyComplianceKPIs: async (
    lineId: string,
    startDate: string,
    endDate: string
  ): Promise<DailyComplianceKPI[]> => {
    const params = new URLSearchParams({
      start_date: startDate,
      end_date: endDate,
    });

    const response = await client.get<DailyComplianceKPI[]>(`/analytics/lines/${lineId}/compliance/daily?${params.toString()}`);
    return response.data;
  },
};

File: api/internal/service/compliance.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
	"go.uber.org/zap"

	"ping/production-line-api/internal/domain"
	"ping/production-line-api/internal/repository"
)

// ComplianceService handles business logic for schedule compliance
type ComplianceService struct {
	complianceRepo *repository.ComplianceRepository
	lineRepo       *repository.LineRepository
	logger         *zap.Logger
}

// NewComplianceService creates a new ComplianceService
func NewComplianceService(
	complianceRepo *repository.ComplianceRepository,
	lineRepo *repository.LineRepository,
	logger *zap.Logger,
) *ComplianceService {
	return &ComplianceService{
		complianceRepo: complianceRepo,
		lineRepo:       lineRepo,
		logger:         logger,
	}
}

// GetAggregateCompliance returns aggregate compliance metrics
func (s *ComplianceService) GetAggregateCompliance(
	ctx context.Context,
	query domain.ComplianceQuery,
) (*domain.AggregateComplianceMetrics, error) {
	// Validate date format
	if err := s.validateDateRange(query.StartDate, query.EndDate); err != nil {
		return nil, err
	}

	s.logger.Info("calculating aggregate compliance",
		zap.String("start_date", query.StartDate),
		zap.String("end_date", query.EndDate),
		zap.Int("line_count", len(query.LineIDs)),
		zap.Int("label_count", len(query.LabelIDs)),
	)

	metrics, err := s.complianceRepo.GetAggregateComplianceMetrics(ctx, query)
	if err != nil {
		s.logger.Error("failed to get aggregate compliance", zap.Error(err))
		return nil, err
	}

	return metrics, nil
}

// GetLineCompliance returns compliance metrics for specific lines
func (s *ComplianceService) GetLineCompliance(
	ctx context.Context,
	query domain.ComplianceQuery,
) ([]domain.LineComplianceMetrics, error) {
	// Validate date format
	if err := s.validateDateRange(query.StartDate, query.EndDate); err != nil {
		return nil, err
	}

	s.logger.Info("calculating line compliance",
		zap.String("start_date", query.StartDate),
		zap.String("end_date", query.EndDate),
	)

	metrics, err := s.complianceRepo.GetLineComplianceMetrics(ctx, query)
	if err != nil {
		s.logger.Error("failed to get line compliance", zap.Error(err))
		return nil, err
	}

	return metrics, nil
}

// GetDailyComplianceKPIs returns daily compliance data for a specific line
func (s *ComplianceService) GetDailyComplianceKPIs(
	ctx context.Context,
	lineID uuid.UUID,
	startDate, endDate string,
) ([]domain.DailyComplianceKPI, error) {
	// Validate date format
	if err := s.validateDateRange(startDate, endDate); err != nil {
		return nil, err
	}

	// Verify line exists
	_, err := s.lineRepo.GetByID(ctx, lineID)
	if err != nil {
		return nil, err
	}

	s.logger.Info("calculating daily compliance KPIs",
		zap.String("line_id", lineID.String()),
		zap.String("start_date", startDate),
		zap.String("end_date", endDate),
	)

	kpis, err := s.complianceRepo.GetDailyComplianceKPIs(ctx, lineID, startDate, endDate)
	if err != nil {
		s.logger.Error("failed to get daily compliance KPIs", zap.Error(err))
		return nil, err
	}

	return kpis, nil
}

// validateDateRange validates that dates are in correct format and range is valid
func (s *ComplianceService) validateDateRange(startDate, endDate string) error {
	start, err := time.Parse("2006-01-02", startDate)
	if err != nil {
		return fmt.Errorf("invalid start_date format, expected YYYY-MM-DD: %w", err)
	}

	end, err := time.Parse("2006-01-02", endDate)
	if err != nil {
		return fmt.Errorf("invalid end_date format, expected YYYY-MM-DD: %w", err)
	}

	if end.Before(start) {
		return fmt.Errorf("end_date must be on or after start_date")
	}

	// Limit range to 90 days to prevent excessive computation
	if end.Sub(start) > 90*24*time.Hour {
		return fmt.Errorf("date range cannot exceed 90 days")
	}

	return nil
}
